craft_application/services/provider.py

```diff
@@ -0,0 +1,181 @@
+    import pathlib
+    from collections.abc import Generator
+
+    import craft_providers
+
+    from craft_application import models
+    from craft_application.app import AppMetadata
+
+
+class ProviderService(base.BaseService):
+    """Manager for craft_providers in an application.
+
+    :param app: Metadata about this application.
+    :param project: The project model
+    :param install_snap: Whether to install this app's snap from the host (default True)
+    """
+
+    managed_mode_env_var = "CRAFT_MANAGED_MODE"
+
+    def __init__(
+        self,
+        app: AppMetadata,
+        project: models.Project,
+        *,
+        install_snap: bool = True,
+    ) -> None:
+        super().__init__(app, project)
+        self._provider: craft_providers.Provider | None = None
+        self.snaps: list[Snap] = []
+        if install_snap:
+            self.snaps.append(Snap(name=app.name, channel=None, classic=True))
+        self.environment: dict[str, str | None] = {self.managed_mode_env_var: "1"}
+
+    @functools.cached_property
+    def is_managed(self) -> bool:
+        """Determine whether we're running in managed mode."""
+        return os.getenv(self.managed_mode_env_var) == "1"
+
+    @contextlib.contextmanager
+    def instance(
+        self,
+        base_name: bases.BaseName | tuple[str, str],
+        *,
+        work_dir: pathlib.Path,
+        allow_unstable: bool = True,
+        **kwargs: bool | str | None,
+    ) -> Generator[craft_providers.Executor, None, None]:
+        """Context manager for getting a provider instance.
+
+        :param base_name: A craft_providers capable base name (tuple of name, version)
+        :param work_dir: Local path to mount inside the provider instance.
+        :param allow_unstable: Whether to allow the use of unstable images.
+        :returns: a context manager of the provider instance.
+        """
+        emit.trace("Preparing managed instance")
+        work_dir_inode = work_dir.stat().st_ino
+        instance_name = f"{self._app.name}-{self._project.name}-{work_dir_inode}"
+        base = self.get_base(base_name, instance_name=instance_name, **kwargs)
+        provider = self.get_provider()
+
+        emit.progress(f"Launching managed {base_name[0]} {base_name[1]} instance...")
+        with provider.launched_environment(
+            project_name=self._project.name,
+            project_path=work_dir,
+            instance_name=instance_name,
+            base_configuration=base,
+            allow_unstable=allow_unstable,
+        ) as instance:
+            instance.mount(
+                host_source=work_dir,
+                # Ignore argument type until craft-providers accepts PurePosixPaths
+                # https://github.com/canonical/craft-providers/issues/315
+                target=self._app.managed_instance_project_path,  # type: ignore[arg-type]
+            )
+            emit.trace("Instance launched and working directory mounted")
```

Comment:
Same here, considering that trace would generate a lot of (potentially machine-generated) output.

Reply:
The thought here was that debug-level logs should be things the app developer should care about and trace would be things that only someone working directly on `craft-application` would care about, but thinking again about this item the app developer probably does care about this. The "Preparing managed instance" message above I could go either way on, but I've deferred to your preference since you probably have a better idea than I do about whether the app dev would care about this.